### PR TITLE
SSO: switch from module codebase to Connection package, part 3

### PR DESCRIPTION
**Pull request by jeherve**
Fixes https://github.com/Automattic/vulcan/issues/274
Fixes https://github.com/Automattic/vulcan/issues/254

## Proposed changes:

This PR is the final part of moving the SSO feature codebase to the Connection package.

It follows the work done in #36587, #36924, and #36989.

- We switch to loading the feature from the connection package in all parts of the Jetpack plugin codebase.
- We deprecate all module codebase classes and methods

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

https://github.com/Automattic/vulcan/issues/251

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

* Go to Jetpack > Settings > Security and enable the SSO feature.
* Log out and log back in, with your main user and with new users.
* Check your error logs and ensure that no deprecation notices appear in the logs.
* Check that there are no errors in CI.

### CODE REVIEW

File: projects/plugins/jetpack/class.jetpack.php
Comment (jeherve) [inline projects/plugins/jetpack/class.jetpack.php]: I'm wondering if I should add a module active check here.

File: projects/plugins/jetpack/modules/sso.php
Comment (jeherve) [inline projects/plugins/jetpack/modules/sso.php:23]: This is mostly only useful to show a configure link in the old module list:

image

File: projects/plugins/jetpack/class.jetpack.php
Comment (fgiannar) [reply projects/plugins/jetpack/class.jetpack.php]: Adding a module active check makes sense to me, so that we only sync those callables when the SSO module is active. 
I would also suggest moving this logic to the Sync package since those callables are no longer Jetpack plugin specific, but available to all plugins via the Connection package. 
Maybe we should also consider adding them to the [minimal Sync configuration as required ones](https://github.com/Automattic/jetpack/tree/trunk/projects/packages/sync#jetpack_sync_callable_whitelist--jetpack_sync_multisite_callable_whitelist) so that if the SSO module is enabled, they will always get synced no matter the consumer Sync configuration. The list of required callables for Sync is defined [here](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/sync/src/class-data-settings.php#L26).
What do you think?

Comment (fgiannar) [reply projects/plugins/jetpack/class.jetpack.php]: Btw I think it's also fine to take care of this in a follow up PR so that we don't introduce too many changes in this one.

Comment (jeherve) [reply projects/plugins/jetpack/class.jetpack.php]: Thanks for the feedback!

> I would also suggest moving this logic to the Sync package since those callables are no longer Jetpack plugin specific, but available to all plugins via the Connection package.

That makes sense. Wouldn't it be better in the Connection package directly, in the SSO class? This way the SSO module check wouldn't be necessary, and the Sync settings would be close to what it relies on / is used for.

> Maybe we should also consider adding them to the [minimal Sync configuration as required ones](https://github.com/Automattic/jetpack/tree/trunk/projects/packages/sync#jetpack_sync_callable_whitelist--jetpack_sync_multisite_callable_whitelist) so that if the SSO module is enabled, they will always get synced no matter the consumer Sync configuration. The list of required callables for Sync is defined [here](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/sync/src/class-data-settings.php#L26).

That's a good point. At the same time, we already have one standalone (A4A) that ships without My Jetpack and is not interested in SSO as far as I know. Wouldn't we want to keep the minimum sync as small as possible for them?

Comment (fgiannar) [reply projects/plugins/jetpack/class.jetpack.php]: > Wouldn't it be better in the Connection package directly, in the SSO class? 

Actually yes, I think that would be the best approach 👍 And since we rely on a filter, it would be fine even in cases where we do have the Connection package but not Sync (eg WooCommerce).

> Wouldn't we want to keep the minimum sync as small as possible for them?

If we add those in the Connection package using the corresponding filter, adding those to Sync's minimal config is not applicable anymore. But even if we decided to have them in Sync, we would only add them to its minimal config when the SSO module is enabled so I believe this wouldn't be a concern for plugins like A4A

Overall I think your suggestion to add those callables in the Connection package is the right one!

Comment (jeherve) [reply projects/plugins/jetpack/class.jetpack.php]: 👍 Done in d05c401a6935cd8cf2477bdbf919ddc753c744a8

File: projects/plugins/jetpack/json-endpoints/jetpack/class.jetpack-json-api-user-create-endpoint.php
Comment (sergeymitr) [inline projects/plugins/jetpack/json-endpoints/jetpack/class.jetpack-json-api-user-create-endpoint.php]: From the looks of it, the endpoint is used for non-SSO related purposes.
It might make sense to extract it from the `SSO\Helpers` while we're at it.

File: projects/plugins/jetpack/modules/sso.php
Comment (sergeymitr) [inline projects/plugins/jetpack/modules/sso.php:31]: Maybe we should preserve the class for now for backward compatibility, something like this:
```
class Jetpack_SSO extends \Automattic\Jetpack\Connection\SSO {}
```

File: projects/plugins/jetpack/modules/sso/class-jetpack-force-2fa.php
Comment (sergeymitr) [inline projects/plugins/jetpack/modules/sso/class-jetpack-force-2fa.php:19]: We probably could remove `force_2fa()` and `get_login_error_message()` as they're private, so we can be sure they aren't used anymore.

File: projects/plugins/jetpack/class.jetpack.php
Comment (jeherve) [reply projects/plugins/jetpack/class.jetpack.php]: hm, the Sync tests are now failing in the Jetpack plugin. Which makes sense since we now hook only when the module is active.
I suppose we could fix the tests by activating the module first, but I wonder if the tests should even stay in the Jetpack plugin, when everything else is moving out of the plugin. What's your take on this? I don't suppose I could run similar tests inside the Connection package, since Sync is not going to be easily accessible there?

File: projects/plugins/jetpack/modules/sso/class.jetpack-sso-user-admin.php
Comment (sergeymitr) [inline projects/plugins/jetpack/modules/sso/class.jetpack-sso-user-admin.php:18]: I think we can remove private properties and methods from this class.

File: projects/plugins/jetpack/tests/php/modules/sso/test_class.jetpack-sso-helpers.php
Comment (sergeymitr) [inline projects/plugins/jetpack/tests/php/modules/sso/test_class.jetpack-sso-helpers.php:9]: Are these tests no longer needed, or we could reuse them for the new `Helpers` class?

File: projects/plugins/jetpack/json-endpoints/jetpack/class.jetpack-json-api-user-create-endpoint.php
Comment (jeherve) [reply projects/plugins/jetpack/json-endpoints/jetpack/class.jetpack-json-api-user-create-endpoint.php]: You make a good point. I'm not sure where I would extract it though, so it's available in both the Jetpack plugin and the Connection package. Do you have any suggestions? Should it be in `Automattic\Jetpack\Connection\Utils`? It feels off since it's not strictly connection-related, but at the same time it does create a user with attached WordPress.com info.

File: projects/plugins/jetpack/modules/sso.php
Comment (jeherve) [reply projects/plugins/jetpack/modules/sso.php:31]: Ah yes, I should have thought about that. Done in c2294edddab0e0dff268c059c9fcd08123a79129

File: projects/plugins/jetpack/modules/sso/class-jetpack-force-2fa.php
Comment (jeherve) [reply projects/plugins/jetpack/modules/sso/class-jetpack-force-2fa.php:19]: Good point. 4e3492feb5fcbee78063965d6c5148a746038196

File: projects/plugins/jetpack/modules/sso/class.jetpack-sso-user-admin.php
Comment (jeherve) [reply projects/plugins/jetpack/modules/sso/class.jetpack-sso-user-admin.php:18]: Done in 7745ce88c3591c0c5a9c89d3787803cd01355aaa

File: projects/plugins/jetpack/tests/php/modules/sso/test_class.jetpack-sso-helpers.php
Comment (jeherve) [reply projects/plugins/jetpack/tests/php/modules/sso/test_class.jetpack-sso-helpers.php:9]: I moved them over to the new class in #36587.

They now live at https://github.com/Automattic/jetpack/blob/6438a55e74ce512ad33b10f34e5b92e9b368c010/projects/packages/connection/tests/php/sso/test_Helpers.php#L1-L15

File: projects/plugins/jetpack/json-endpoints/jetpack/class.jetpack-json-api-user-create-endpoint.php
Comment (sergeymitr) [reply projects/plugins/jetpack/json-endpoints/jetpack/class.jetpack-json-api-user-create-endpoint.php]: Yeah, it does look out of place wherever we put it.
I can't think of anything better than `Utils`.

Comment (jeherve) [reply projects/plugins/jetpack/json-endpoints/jetpack/class.jetpack-json-api-user-create-endpoint.php]: Done in d1571de8f8424a76bded6eabb26410826eea3745

File: projects/plugins/jetpack/class.jetpack.php
Comment (jeherve) [reply projects/plugins/jetpack/class.jetpack.php]: @fgiannar How do you think I should address those changes to the Sync tests? Do you have any ideas, by any chance?

Comment (fgiannar) [reply projects/plugins/jetpack/class.jetpack.php]: Given the nature of the sync tests we run in the Jetpack plugin (integration vs unit) I think it's best to leave them inside the Jetpack plugin (we even do the same for Woo related ones) and activate the module first as you suggested.

Btw apologies for the delayed reply here as I missed this one coming back from AFK.

**Comment by jeherve** [comment]
@sergeymitr @fgiannar Thank you both for the extensive reviews so far! This should be good for a (hopefully final) review!

File: projects/packages/connection/src/sso/class-helpers.php
Comment (sergeymitr) [inline projects/packages/connection/src/sso/class-helpers.php:222]: We added the method very recently, but it's already been released.
Do you think it's worth the trouble to deprecate it first?

**Review by sergeymitr — DISMISSED** [review]

**Comment by sergeymitr** [comment]
Looks good and works well 👍 
The `generate_user()` thing is not a blocker, whatever you decide to do with it.
I also didn't test the Sync changes, hopefully @fgiannar can take a look.

**Comment by fgiannar** [comment]
Unfortunately the SSO callables are not synced as expected, however this can be fixed in general in a separate PR:
It happens because of [the module code running on late_initialization](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/class.jetpack.php#L696) and the sync callable whitelist being also set on `plugins_loaded` with the same priority 90 as configured [here](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/sync/src/class-main.php#L26).

I'm currently working on a fix, hopefully we can resume testing once it's merged :)

**Review by fgiannar — APPROVED** [review]
LGTM! I can confirm that after merging #37370 sso callables are now synced as expected.

I also tested tweaking the SSO settings and logging in and out with an admin and an editor (which I invited) and everything works as expected. 
Didn't see any deprecation notices in the logs or related errors in CI 👍

**Pull request merged**
